File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/graph/P207_CourseSchedule1.py
from collections import deque
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph, in_degrees = self.build_graph_and_in_degrees(numCourses, prerequisites)

        queue = deque()
        for i in range(numCourses):
            if in_degrees[i] == 0:
                queue.append(i)
        while queue:
            i = queue.popleft()
            numCourses -= 1
            for _to in graph[i]:
                in_degrees[_to] -= 1
                if in_degrees[_to] == 0:
                    queue.append(_to)
        return numCourses == 0

    def build_graph_and_in_degrees(self, numCourses, prerequisites):
        graph = [[] for _ in range(numCourses)]
        in_degrees = [0] * numCourses
        for edge in prerequisites:
            _to = edge[0]
            _from = edge[1]
            graph[_from].append(_to)
            in_degrees[_to] += 1
        return graph, in_degrees

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/graph/P207_CourseSchedule1.java
package com.dsa.leetcode.labuladong.graph;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @since 2024/6/28 22:31
 */
public class P207_CourseSchedule1 {


    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = buildGraph(numCourses, prerequisites);
        int[] inDegrees = buildInDegrees(numCourses, prerequisites);

        // find courses with zero inDegree
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegrees[i] == 0) {
                queue.add(i);
            }
        }

        while (!queue.isEmpty()) {
            Integer from = queue.poll();
            numCourses--;
            for (Integer to : graph.get(from)) {
                inDegrees[to] -= 1;
                if (inDegrees[to] == 0) {
                    queue.offer(to);
                }
            }
        }
        return numCourses == 0;
    }

    private int[] buildInDegrees(int numCourses, int[][] prerequisites) {
        int[] inDegrees = new int[numCourses];
        for (int[] edge : prerequisites) {
            int to = edge[0];
            inDegrees[to] += 1;
        }
        return inDegrees;
    }

    private List<List<Integer>> buildGraph(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>(numCourses);
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : prerequisites) {
            int to = edge[0];
            int from = edge[1];
            graph.get(from).add(to);
        }
        return graph;
    }


    public static void main(String[] args) {
        int numCourse = 2;
        int[][] prerequisites = {{1, 0}, {0, 1}};
        P207_CourseSchedule1 p207 = new P207_CourseSchedule1();
        boolean b = p207.canFinish(numCourse, prerequisites);
        System.out.println(b);

    }
}
